File: readers/DelphesProMC.cpp
/*
 *  Delphes: a framework for fast simulation of a generic collider experiment
 *  Copyright (C) 2012-2014  Universite catholique de Louvain (UCL), Belgium
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdexcept>
#include <iostream>
#include <sstream>
#include <memory>

#include <map>

#include <stdlib.h>
#include <signal.h>
#include <stdio.h>

#include "TROOT.h"
#include "TApplication.h"

#include "TFile.h"
#include "TObjArray.h"
#include "TStopwatch.h"
#include "TDatabasePDG.h"
#include "TParticlePDG.h"
#include "TLorentzVector.h"

#include "modules/Delphes.h"
#include "classes/DelphesStream.h"
#include "classes/DelphesClasses.h"
#include "classes/DelphesFactory.h"

#include "ExRootAnalysis/ExRootTreeWriter.h"
#include "ExRootAnalysis/ExRootTreeBranch.h"
#include "ExRootAnalysis/ExRootProgressBar.h"

#include "ProMC.pb.h"
#include "ProMCBook.h"
#include "ProMCHeader.pb.h"

using namespace std;

//---------------------------------------------------------------------------

void ConvertInput(ProMCEvent &event, double momentumUnit, double positionUnit,
  ExRootTreeBranch *branch, DelphesFactory *factory,
  TObjArray *allParticleOutputArray, TObjArray *stableParticleOutputArray,
  TObjArray *partonOutputArray, TStopwatch *readStopWatch, TStopwatch *procStopWatch)
{
  Int_t i;

  ProMCEvent_Event *mutableEvent;
  ProMCEvent_Particles *mutableParticles;

  HepMCEvent *element;
  Candidate *candidate;
  TDatabasePDG *pdg;
  TParticlePDG *pdgParticle;
  Int_t pdgCode;

  Int_t pid, status;
  Double_t px, py, pz, mass;
  Double_t x, y, z, t;

  pdg = TDatabasePDG::Instance();

  // event information
  mutableEvent = event.mutable_event();

  element = static_cast<HepMCEvent *>(branch->NewEntry());

  element->Number = mutableEvent->number();

  element->ProcessID = mutableEvent->process_id();
  element->MPI = mutableEvent->mpi();
  element->Weight = mutableEvent->weight();
  element->Scale = mutableEvent->scale();
  element->AlphaQED = mutableEvent->alpha_qed();
  element->AlphaQCD = mutableEvent->alpha_qcd();

  element->ID1 = mutableEvent->id1();
  element->ID2 = mutableEvent->id2();
  element->X1 = mutableEvent->x1();
  element->X2 = mutableEvent->x2();
  element->ScalePDF = mutableEvent->scale_pdf();
  element->PDF1 = mutableEvent->pdf1();
  element->PDF2 = mutableEvent->pdf2();

  element->ReadTime = readStopWatch->RealTime();
  element->ProcTime = procStopWatch->RealTime();

  mutableParticles = event.mutable_particles();

  for(i = 0; i < mutableParticles->pdg_id_size(); ++i)
  {
    pid = mutableParticles->pdg_id(i);
    status = mutableParticles->status(i);

    px = mutableParticles->px(i)/momentumUnit;
    py = mutableParticles->py(i)/momentumUnit;
    pz = mutableParticles->pz(i)/momentumUnit;
    mass = mutableParticles->mass(i)/momentumUnit;
    x = mutableParticles->x(i)/positionUnit;
    y = mutableParticles->y(i)/positionUnit;
    z = mutableParticles->z(i)/positionUnit;
    t = mutableParticles->t(i)/positionUnit;

    candidate = factory->NewCandidate();

    candidate->PID = pid;
    pdgCode = TMath::Abs(candidate->PID);

    candidate->Status = status;

    candidate->M1 = mutableParticles->mother1(i);
    candidate->M2 = mutableParticles->mother2(i);

    candidate->D1 = mutableParticles->daughter1(i);
    candidate->D2 = mutableParticles->daughter2(i);

    pdgParticle = pdg->GetParticle(pid);
    candidate->Charge = pdgParticle ? Int_t(pdgParticle->Charge()/3.0) : -999;
    candidate->Mass = mass;

    candidate->Momentum.SetXYZM(px, py, pz, mass);

    candidate->Position.SetXYZT(x, y, z, t);

    allParticleOutputArray->Add(candidate);

    if(!pdgParticle) continue;

    if(status == 1)
    {
      stableParticleOutputArray->Add(candidate);
    }
    else if(pdgCode <= 5 || pdgCode == 21 || pdgCode == 15)
    {
      partonOutputArray->Add(candidate);
    }
  }
}

//---------------------------------------------------------------------------

static bool interrupted = false;

void SignalHandler(int sig)
{
  interrupted = true;
}

//---------------------------------------------------------------------------

int main(int argc, char *argv[])
{
  char appName[] = "DelphesProMC";
  stringstream message;
  ProMCBook *inputFile = 0;
  TFile *outputFile = 0;
  TStopwatch readStopWatch, procStopWatch;
  ExRootTreeWriter *treeWriter = 0;
  ExRootTreeBranch *branchEvent = 0;
  ExRootConfReader *confReader = 0;
  Delphes *modularDelphes = 0;
  DelphesFactory *factory = 0;
  TObjArray *allParticleOutputArray = 0, *stableParticleOutputArray = 0, *partonOutputArray = 0;
  Int_t i;
  Long64_t eventCounter, numberOfEvents;
  double momentumUnit = 1.0, positionUnit = 1.0;

  if(argc < 4)
  {
    cout << " Usage: " << appName << " config_file" << " output_file" << " input_file(s)" << endl;
    cout << " config_file - configuration file in Tcl format," << endl;
    cout << " output_file - output file in ROOT format," << endl;
    cout << " input_file(s) - input file(s) in ProMC format." << endl;
    return 1;
  }

  signal(SIGINT, SignalHandler);

  gROOT->SetBatch();

  int appargc = 1;
  char *appargv[] = {appName};
  TApplication app(appName, &appargc, appargv);

  try
  {
    outputFile = TFile::Open(argv[2], "CREATE");

    if(outputFile == NULL)
    {
      message << "can't open " << argv[2] << endl;
      throw runtime_error(message.str());
    }

    treeWriter = new ExRootTreeWriter(outputFile, "Delphes");

    branchEvent = treeWriter->NewBranch("Event", HepMCEvent::Class());

    confReader = new ExRootConfReader;
    confReader->ReadFile(argv[1]);

    modularDelphes = new Delphes("Delphes");
    modularDelphes->SetConfReader(confReader);
    modularDelphes->SetTreeWriter(treeWriter);

    factory = modularDelphes->GetFactory();
    allParticleOutputArray = modularDelphes->ExportArray("allParticles");
    stableParticleOutputArray = modularDelphes->ExportArray("stableParticles");
    partonOutputArray = modularDelphes->ExportArray("partons");

    modularDelphes->InitTask();

    for(i = 3; i < argc && !interrupted; ++i)
    {
      cout << "** Reading " << argv[i] << endl;

      inputFile = new ProMCBook(argv[i], "r");

      ProMCHeader header = inputFile->getHeader();

      momentumUnit = static_cast<double>(header.momentumunit());
      positionUnit = static_cast<double>(header.lengthunit());



      if(inputFile == NULL)
      {
        message << "can't open " << argv[i] << endl;
        throw runtime_error(message.str());
      }

      numberOfEvents = inputFile->getEvents();

      if(numberOfEvents <= 0) continue;

      ExRootProgressBar progressBar(numberOfEvents - 1);

      // Loop over all objects
      modularDelphes->Clear();
      treeWriter->Clear();
      readStopWatch.Start();
      for(eventCounter = 0; eventCounter < numberOfEvents && !interrupted; ++eventCounter)
      {
        if(inputFile->next() != 0) continue;
        ProMCEvent event = inputFile->get();

        readStopWatch.Stop();

        procStopWatch.Start();
        ConvertInput(event, momentumUnit, positionUnit,
          branchEvent, factory,
          allParticleOutputArray, stableParticleOutputArray,
          partonOutputArray, &readStopWatch, &procStopWatch);
        modularDelphes->ProcessTask();
        procStopWatch.Stop();

        treeWriter->Fill();

        modularDelphes->Clear();
        treeWriter->Clear();

        readStopWatch.Start();
        progressBar.Update(eventCounter);
      }

      progressBar.Update(eventCounter, eventCounter, kTRUE);
      progressBar.Finish();

      inputFile->close();
      delete inputFile;
    }

    modularDelphes->FinishTask();
    treeWriter->Write();

    cout << "** Exiting..." << endl;

    delete modularDelphes;
    delete confReader;
    delete treeWriter;
    delete outputFile;

    return 0;
  }
  catch(runtime_error &e)
  {
    if(treeWriter) delete treeWriter;
    if(outputFile) delete outputFile;
    cerr << "** ERROR: " << e.what() << endl;
    return 1;
  }
}

File: readers/DelphesProMC.py
#!/usr/bin/env python3

import sys

from urllib.request import Request, urlopen
import zipfile

import ROOT

import ProMCHeader_pb2
import ProMC_pb2

################################################################################
# HttpFile class is written by Eli Carter (retracile)
# http://stackoverflow.com/a/7852229


class HttpFile(object):
    def __init__(self, url):
        self.url = url
        self.offset = 0
        self.length = -1

    def seekable(self):
        return True

    def size(self):
        if self.length < 0:
            f = urlopen(self.url)
            self.length = int(f.headers["Content-length"])
        return self.length

    def read(self, count=-1):
        req = Request(self.url)
        if count < 0:
            end = self.size() - 1
        else:
            end = self.offset + count - 1
        req.headers["Range"] = "bytes=%s-%s" % (self.offset, end)
        f = urlopen(req)
        data = f.read()
        # FIXME: should check that we got the range expected, etc.
        chunk = len(data)
        if count >= 0:
            assert chunk == count
        self.offset += chunk
        return data

    def seek(self, offset, whence=0):
        if whence == 0:
            self.offset = offset
        elif whence == 1:
            self.offset += offset
        elif whence == 2:
            self.offset = self.size() + offset
        else:
            raise Exception("Invalid whence")

    def tell(self):
        return self.offset


################################################################################


def ConvertInput(input, momentumUnit, lengthUnit, branch, factory, allParticleOutputArray, stableParticleOutputArray, partonOutputArray):

    pdg = ROOT.TDatabasePDG.Instance()

    data = ProMC_pb2.ProMCEvent()
    data.ParseFromString(input)

    event = data.event
    particles = data.particles

    element = branch.NewEntry()

    element.Number = event.Number if event.Number else 0
    element.ProcessID = event.Process_ID if event.Process_ID else 0
    element.MPI = event.MPI if event.MPI else 0
    element.Weight = event.Weight if event.Weight else 0
    element.Scale = event.Scale if event.Scale else 0
    element.AlphaQED = event.Alpha_QED if event.Alpha_QED else 0
    element.AlphaQCD = event.Alpha_QCD if event.Alpha_QCD else 0
    element.ID1 = event.ID1 if event.ID1 else 0
    element.ID2 = event.ID2 if event.ID2 else 0
    element.X1 = event.X1 if event.X1 else 0
    element.X2 = event.X2 if event.X2 else 0
    element.ScalePDF = event.Scale_PDF if event.Scale_PDF else 0
    element.PDF1 = event.PDF1 if event.PDF1 else 0
    element.PDF2 = event.PDF2 if event.PDF2 else 0

    for i in range(len(particles.pdg_id)):
        pid = particles.pdg_id[i]
        status = particles.status[i]
        px = particles.Px[i]
        py = particles.Py[i]
        pz = particles.Pz[i]
        mass = particles.mass[i]

        x = particles.X[i]
        y = particles.Y[i]
        z = particles.Z[i]
        t = particles.T[i]

        candidate = factory.NewCandidate()

        candidate.PID = pid

        candidate.Status = status

        candidate.M1 = particles.mother1[i] if particles.mother1[i] < sys.maxsize else -1
        candidate.M2 = particles.mother2[i] if particles.mother2[i] < sys.maxsize else -1

        candidate.D1 = particles.daughter1[i] if particles.daughter1[i] < sys.maxsize else -1
        candidate.D2 = particles.daughter2[i] if particles.daughter2[i] < sys.maxsize else -1

        candidate.Mass = mass

        candidate.Momentum.SetXYZM(px, py, pz, mass)

        candidate.Position.SetXYZT(x, y, z, t)

        pdgParticle = pdg.GetParticle(pid)

        candidate.Charge = int(pdgParticle.Charge() / 3.0) if pdgParticle else -999

        allParticleOutputArray.Add(candidate)

        if not pdgParticle:
            continue

        pdgCode = abs(pid)

        if status == 1:
            stableParticleOutputArray.Add(candidate)
        elif pdgCode <= 5 or pdgCode == 21 or pdgCode == 15:
            partonOutputArray.Add(candidate)


################################################################################

if len(sys.argv) < 4:
    print(" Usage: DelphesProMC.py config_file output_file input_file(s)")
    sys.exit(1)

ROOT.gSystem.Load("libDelphes")

outputFile = ROOT.TFile(sys.argv[2], "CREATE")

if not outputFile.IsOpen():
    print("** ERROR: can't open", sys.argv[2])
    sys.exit(1)

treeWriter = ROOT.ExRootTreeWriter(outputFile, "Delphes")

branchEvent = treeWriter.NewBranch("Event", ROOT.HepMCEvent.Class())

confReader = ROOT.ExRootConfReader()
confReader.ReadFile(sys.argv[1])

modularDelphes = ROOT.Delphes("Delphes")
modularDelphes.SetConfReader(confReader)
modularDelphes.SetTreeWriter(treeWriter)

factory = modularDelphes.GetFactory()
allParticleOutputArray = modularDelphes.ExportArray("allParticles")
stableParticleOutputArray = modularDelphes.ExportArray("stableParticles")
partonOutputArray = modularDelphes.ExportArray("partons")

modularDelphes.InitTask()

for fileName in sys.argv[3:]:
    print("** Reading", fileName)

    if fileName.startswith("http://") or fileName.startswith("https://"):
        file = HttpFile(fileName)
    else:
        file = fileName

    with zipfile.ZipFile(file) as zf:

        numberOfEvents = len(zf.namelist())
        if numberOfEvents <= 0:
            continue

        progressBar = ROOT.ExRootProgressBar(numberOfEvents - 1)

        # retrive information from the header file
        header = ProMCHeader_pb2.ProMCHeader()
        header.ParseFromString(zf.read("header"))
        momentumUnit = float(header.MomentumUnit)
        lengthUnit = float(header.LengthUnit)

        modularDelphes.Clear()
        treeWriter.Clear()
        eventCounter = 0
        for name in zf.namelist():
            eventCounter += 1
            if not name.isdigit():
                continue

            ConvertInput(zf.read(name), momentumUnit, lengthUnit, branchEvent, factory, allParticleOutputArray, stableParticleOutputArray, partonOutputArray)

            modularDelphes.ProcessTask()

            treeWriter.Fill()

            modularDelphes.Clear()
            treeWriter.Clear()
            progressBar.Update(eventCounter)

        progressBar.Update(eventCounter, eventCounter, ROOT.kTRUE)
        progressBar.Finish()

modularDelphes.FinishTask()
treeWriter.Write()

del treeWriter
